### PR TITLE
Extract block announce validation from `ChainSync`

### PR DESCRIPTION
**Pull request by dmitry-markin**
This PR is part of Sync 2.0 preparation work. Resolves https://github.com/paritytech/polkadot-sdk/issues/503.

Supersedes https://github.com/paritytech/substrate/pull/14675, please see the discussion there for details.

### CODE REVIEW

File: substrate/client/network/sync/src/futures_stream.rs [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+use std::{
+	pin::Pin,
+	task::{Context, Poll, Waker},
+};
+
+/// Wrapper around [`FuturesUnordered`] that wakes a task up automatically.
+pub struct FuturesStream<F> {
+	futures: FuturesUnordered<F>,
+	waker: Option<Waker>,
+}
```
Comment (dmitry-markin) [inline substrate/client/network/sync/src/futures_stream.rs:32]: @altonen I've replaced the channel-based version of `FuturesStream` with a `Waker`-based in this PR. We can roll things back later if needed.

**Review by altonen — APPROVED** [review]

File: substrate/client/network/sync/src/futures_stream.rs [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+	pub fn len(&self) -> usize {
+		self.futures.len()
+	}
+}
+
+impl<F: Future> Stream for FuturesStream<F> {
+	type Item = <F as Future>::Output;
+
+	fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+		if let Poll::Ready(Some(result)) = self.futures.poll_next_unpin(cx) {
```
[the remark this answers is not in the document]
Comment (dmitry-markin) [reply substrate/client/network/sync/src/futures_stream.rs]: Done

**Review by kpp — APPROVED** [review]

File: substrate/client/network/sync/src/lib.rs [hunk 3]
```diff
@@ -1107,119 +1040,88 @@ where
...
-		let future = self.block_announce_validator.validate(header, assoc_data);
+		// If the announced block is the best they have and is not ahead of us, our common number
+		// is either one further ahead or it's the one they just announced, if we know about it.
+		if is_best {
+			if known && self.best_queued_number >= number {
+				self.update_peer_common_number(&who, number);
+			} else if announce.header.parent_hash() == &self.best_queued_hash ||
+				known_parent && self.best_queued_number >= number
+			{
+				self.update_peer_common_number(&who, number - One::one());
```
Comment (kpp) [inline substrate/client/network/sync/src/lib.rs:1083]: saturating?

Comment (dmitry-markin) [reply substrate/client/network/sync/src/lib.rs]: Done

**Comment by altonen** [comment]
Any reason this isn't merged yet?

**Comment by dmitry-markin** [comment]
> Any reason this isn't merged yet?

Last week one zombienet test was failing (I guess on master too), so I've been waiting for the CI to turn green.

**Comment by dmitry-markin** [comment]
And now **Check Markdown** test is failing :see_no_evil:

**Pull request merged**